File: src/content/docs/topics/cs/design-patterns/state.py
from __future__ import annotations
from typing import Protocol


class TrafficLightState(Protocol):
    def next(self, context: TrafficLight) -> None: ...
    def get_color(self) -> str: ...


class RedState:
    def next(self, context: TrafficLight) -> None:
        context.set_state(GreenState())

    def get_color(self) -> str:
        return 'Red'


class GreenState:
    def next(self, context: TrafficLight) -> None:
        context.set_state(YellowState())

    def get_color(self) -> str:
        return 'Green'


class YellowState:
    def next(self, context: TrafficLight) -> None:
        context.set_state(RedState())

    def get_color(self) -> str:
        return 'Yellow'


class TrafficLight:
    def __init__(self) -> None:
        self._state: TrafficLightState = RedState()

    def set_state(self, state: TrafficLightState) -> None:
        self._state = state

    def next(self) -> None:
        self._state.next(self)

    def get_color(self) -> str:
        return self._state.get_color()


light = TrafficLight()
print(light.get_color())  # Red
light.next()
print(light.get_color())  # Green
light.next()
print(light.get_color())  # Yellow
light.next()
print(light.get_color())  # Red

File: src/content/docs/topics/cs/design-patterns/state.ts
interface TrafficLightState {
  next(context: TrafficLight): void;
  getColor(): string;
}

class RedState implements TrafficLightState {
  next(context: TrafficLight): void {
    context.setState(new GreenState());
  }
  getColor(): string {
    return 'Red';
  }
}

class GreenState implements TrafficLightState {
  next(context: TrafficLight): void {
    context.setState(new YellowState());
  }
  getColor(): string {
    return 'Green';
  }
}

class YellowState implements TrafficLightState {
  next(context: TrafficLight): void {
    context.setState(new RedState());
  }
  getColor(): string {
    return 'Yellow';
  }
}

class TrafficLight {
  constructor(private state: TrafficLightState = new RedState()) {}

  setState(state: TrafficLightState): void {
    this.state = state;
  }

  next(): void {
    this.state.next(this);
  }

  getColor(): string {
    return this.state.getColor();
  }
}

const light = new TrafficLight();
console.log(light.getColor()); // Red
light.next();
console.log(light.getColor()); // Green
light.next();
console.log(light.getColor()); // Yellow
light.next();
console.log(light.getColor()); // Red
